### ISSUE TITLE
Fetch system events

### ISSUE THREAD

**Issue by mariopino** [Colm3na/polkastats-backend-v3#5]
I think we can use the block crawlers to store system events, so there's no need to create separate crawlers.

Needed logic is already done in backend v2:

https://github.com/Colm3na/polkastats-backend-v2/blob/master/crawlers/events.js

**Comment by mariopino** [comment]
This is almost done thought there is a pending task:

- Update counters in block table: total_extrinsics, total_signed_extrinsics, total_failed_extrinsics, total_events, total_system_events, total_module_events, new_accounts, reaped_accounts, new_contracts and new_sessions

**Comment by mariopino** [comment]
Bug detected in Polkadot JS rpc api:

```
const block = 3876;
const blockHash = await api.rpc.chain.getBlockHash(block);
const blockEvents = await api.query.system.events.at(blockHash);
```

Result:

```
Error: getStorage(key: StorageKey, at?: BlockHash): StorageData:: createType(Vec<EventRecord>):: Unable to create Enum via index 5, in ApplyExtrinsic, Finalization
    at e.selector (https://polkadot.js.org/apps/main.5fcc7456.js:65:323920)
    at e.error (https://polkadot.js.org/apps/vendor.af33e3c6.js:1:57594)
    at e._next (https://polkadot.js.org/apps/vendor.af33e3c6.js:8:20912)
    at e.next (https://polkadot.js.org/apps/vendor.af33e3c6.js:1:971)
    at e.notifyNext (https://polkadot.js.org/apps/vendor.af33e3c6.js:1:95724)
    at e._next (https://polkadot.js.org/apps/vendor.af33e3c6.js:8:22338)
    at e.next (https://polkadot.js.org/apps/vendor.af33e3c6.js:1:971)
    at e.notifyNext (https://polkadot.js.org/apps/vendor.af33e3c6.js:8:88172)
    at e._next (https://polkadot.js.org/apps/vendor.af33e3c6.js:8:22338)
    at e.next (https://polkadot.js.org/apps/vendor.af33e3c6.js:1:971)
```

**Comment by mariopino** [comment]
Well, this is not a bug, as @jacogr clarified:

> You cannot query historic events this way
> 
> - the API only decorated the latest metadata
> - the types are outdated at that point
> 
> See https://github.com/polkadot-js/api/blob/master/packages/types/src/known/overrides.ts#L50 & #845
> 
> TL;DR The API is a real-time tool, atm for historical queries, you are on your own. (The types and metadata for historical blocks need to be injected). There may be an "archive mode" at some point, however at this point querying both the state and the metadata at the previous block (metadat takes effect at the end of the block), it out of scope.

Sooo we have a problem! :-D

But for a new network, it's not really a big problem, because you can:

1. Run `listener` when network starts to produce blocks. Also periodically run `harvester` to fill the gaps (as usually).
2. Eventually, substrate runtime will be updated, at that point system events collection may stop (metadata has changed and data extraction probably will fail, listener is not able to read the new runtime metadata to get the events).
3. Stop the crawler, update to latest Polkadot API so you have the updated types.
4. Run `listener` to keep fetching latest blocks and run `harvester` as usual, it should add all the missing events (it haves the latest types so can understand latest runtime metadata).
5. Profit! We have a database with all the information updated!

**Comment by jacogr** [comment]
On 2 - in API 1.4 the metadata is redecorated on upgrades (augmented with new metadata), so assuming you are running a recent API with all the types, e.g. for [Kusama](https://github.com/polkadot-js/api/blob/master/packages/types/src/known/overrides.ts#L53), it will just continue.

**Comment by mariopino** [comment]
> On 2 - in API 1.4 the metadata is redecorated on upgrades (augmented with new metadata), so assuming you are running a recent API with all the types, e.g. for [Kusama](https://github.com/polkadot-js/api/blob/master/packages/types/src/known/overrides.ts#L53), it will just continue.

This is great @jacogr ! Thank you for the update!

**Comment by jacogr** [comment]
Forgot to mention, example here - https://github.com/polkadot-js/api/pull/1896#issue-377040103

When calls are added/removed, it also displays those.

(A bit of a crappy test, but really worst-case - "upgraded" from substrate to polkadot, bit of hacking to ignore genesisHash which is generally checked. But had to try something quite big to make sure the augmentation works and even pulls through in stuff like the apps UI)

**Comment by mariopino** [comment]
Closing this as we meet the first milestone target, the pending task about updating the block counters will be in another issue.
